lib/l1_watcher/src/upgrade_tx_watcher.rs

```diff
@@ -0,0 +1,311 @@
+use crate::L1WatcherConfig;
+use crate::watcher::{L1Watcher, L1WatcherError, ProcessL1Event};
+use alloy::dyn_abi::SolType;
+use alloy::primitives::{Address, B256, U256};
+use alloy::providers::{DynProvider, Provider};
+use alloy::rpc::types::{Filter, Log};
+use alloy::sol_types::SolEvent;
+use tokio::sync::mpsc;
+use zksync_os_contract_interface::IChainAdmin::UpdateUpgradeTimestamp;
+use zksync_os_contract_interface::IChainTypeManager::{NewUpgradeCutData, ProposedUpgrade};
+use zksync_os_contract_interface::ZkChain;
+use zksync_os_types::{L1UpgradeEnvelope, ProtocolSemanticVersion, UpgradeTransaction};
+
+// TODO: disabled until bytecode supplier integration is ready
+// use zksync_os_contract_interface::IBytecodeSupplier::BytecodePublished;
+// use zk_os_api::helpers::set_properties_code;
+// use zk_os_basic_system::system_implementation::flat_storage_model::AccountProperties;
+
+/// Don't try to process that many block linearly
+/// The constant value is higher than for other watchers, since we're looking for rare/specific events
+/// and we don't expect a lot of results.
+const MAX_L1_BLOCKS_LOOKBEHIND: u64 = 2_500_000;
```

Comment:
Top part of the comment doesn't make sense with the current implementation. All other L1 watchers run binsearch to find the starting L1 block (in your case that would be where current protocol version got activated I guess). I know you already expressed your dislike of running binsearch but I feel like the current approach with linear search on last 2.5M blocks on every restart is even more brittle, no?

Reply:
Good point, already working on changing this part.